// css-selector-parser/dist/cjs/syntax-definitions.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cssSyntaxDefinitions = exports.extendSyntaxDefinition = exports.getXmlOptions = void 0;
var emptyXmlOptions = {};
var defaultXmlOptions = { wildcard: true };
function getXmlOptions(param) {
    if (param) {
        if (typeof param === 'boolean') {
            return defaultXmlOptions;
        }
        else {
            return param;
        }
    }
    else {
        return emptyXmlOptions;
    }
}
exports.getXmlOptions = getXmlOptions;
function extendSyntaxDefinition(base, extension) {
    var result = __assign({}, base);
    if ('tag' in extension) {
        if (extension.tag) {
            result.tag = __assign({}, getXmlOptions(base.tag));
            var extensionOptions = getXmlOptions(extension.tag);
            if ('wildcard' in extensionOptions) {
                result.tag.wildcard = extensionOptions.wildcard;
            }
        }
        else {
            result.tag = undefined;
        }
    }
    if ('ids' in extension) {
        result.ids = extension.ids;
    }
    if ('classNames' in extension) {
        result.classNames = extension.classNames;
    }
    if ('namespace' in extension) {
        if (extension.namespace) {
            result.namespace = __assign({}, getXmlOptions(base.namespace));
            var extensionOptions = getXmlOptions(extension.namespace);
            if ('wildcard' in extensionOptions) {
                result.namespace.wildcard = extensionOptions.wildcard;
            }
        }
        else {
            result.namespace = undefined;
        }
    }
    if ('combinators' in extension) {
        if (extension.combinators) {
            result.combinators = result.combinators
                ? result.combinators.concat(extension.combinators)
                : extension.combinators;
        }
        else {
            result.combinators = undefined;
        }
    }
    if ('attributes' in extension) {
        if (extension.attributes) {
            result.attributes = __assign({}, base.attributes);
            if ('unknownCaseSensitivityModifiers' in extension.attributes) {
                result.attributes.unknownCaseSensitivityModifiers =
                    extension.attributes.unknownCaseSensitivityModifiers;
            }
            if ('operators' in extension.attributes) {
                result.attributes.operators = extension.attributes.operators
                    ? result.attributes.operators
                        ? result.attributes.operators.concat(extension.attributes.operators)
                        : extension.attributes.operators
                    : undefined;
            }
            if ('caseSensitivityModifiers' in extension.attributes) {
                result.attributes.caseSensitivityModifiers = extension.attributes.caseSensitivityModifiers
                    ? result.attributes.caseSensitivityModifiers
                        ? result.attributes.caseSensitivityModifiers.concat(extension.attributes.caseSensitivityModifiers)
                        : extension.attributes.caseSensitivityModifiers
                    : undefined;
            }
        }
        else {
            result.attributes = undefined;
        }
    }
    if ('pseudoElements' in extension) {
        if (extension.pseudoElements) {
            result.pseudoElements = __assign({}, base.pseudoElements);
            if ('unknown' in extension.pseudoElements) {
                result.pseudoElements.unknown = extension.pseudoElements.unknown;
            }
            if ('notation' in extension.pseudoElements) {
                result.pseudoElements.notation = extension.pseudoElements.notation;
            }
            if ('definitions' in extension.pseudoElements) {
                result.pseudoElements.definitions = extension.pseudoElements.definitions
                    ? result.pseudoElements.definitions
                        ? result.pseudoElements.definitions.concat(extension.pseudoElements.definitions)
                        : extension.pseudoElements.definitions
                    : undefined;
            }
        }
        else {
            result.pseudoElements = undefined;
        }
    }
    if ('pseudoClasses' in extension) {
        if (extension.pseudoClasses) {
            result.pseudoClasses = __assign({}, base.pseudoClasses);
            if ('unknown' in extension.pseudoClasses) {
                result.pseudoClasses.unknown = extension.pseudoClasses.unknown;
            }
            if ('definitions' in extension.pseudoClasses) {
                var newDefinitions = extension.pseudoClasses.definitions;
                if (newDefinitions) {
                    result.pseudoClasses.definitions = __assign({}, result.pseudoClasses.definitions);
                    var existingDefinitions = result.pseudoClasses.definitions;
                    for (var _i = 0, _a = Object.keys(newDefinitions); _i < _a.length; _i++) {
                        var key = _a[_i];
                        var newDefinitionForNotation = newDefinitions[key];
                        var existingDefinitionForNotation = existingDefinitions[key];
                        if (newDefinitionForNotation) {
                            existingDefinitions[key] = existingDefinitionForNotation
                                ? existingDefinitionForNotation.concat(newDefinitionForNotation)
                                : newDefinitionForNotation;
                        }
                        else {
                            existingDefinitions[key] = undefined;
                        }
                    }
                }
                else {
                    result.pseudoClasses.definitions = undefined;
                }
            }
        }
        else {
            result.pseudoClasses = undefined;
        }
    }
    return result;
}
exports.extendSyntaxDefinition = extendSyntaxDefinition;
var css1SyntaxDefinition = {
    tag: {},
    ids: true,
    classNames: true,
    combinators: [],
    pseudoElements: {
        unknown: 'reject',
        notation: 'singleColon',
        definitions: ['first-letter', 'first-line']
    },
    pseudoClasses: {
        unknown: 'reject',
        definitions: {
            NoArgument: ['link', 'visited', 'active']
        }
    }
};
var css2SyntaxDefinition = extendSyntaxDefinition(css1SyntaxDefinition, {
    tag: { wildcard: true },
    combinators: ['>', '+'],
    attributes: {
        unknownCaseSensitivityModifiers: 'reject',
        operators: ['=', '~=', '|=']
    },
    pseudoElements: {
        definitions: ['before', 'after']
    },
    pseudoClasses: {
        unknown: 'reject',
        definitions: {
            NoArgument: ['hover', 'focus', 'first-child'],
            String: ['lang']
        }
    }
});
var selectors3SyntaxDefinition = extendSyntaxDefinition(css2SyntaxDefinition, {
    namespace: {
        wildcard: true
    },
    combinators: ['~'],
    attributes: {
        operators: ['^=', '$=', '*=']
    },
    pseudoElements: {
        notation: 'both'
    },
    pseudoClasses: {
        definitions: {
            NoArgument: [
                'root',
                'last-child',
                'first-of-type',
                'last-of-type',
                'only-child',
                'only-of-type',
                'empty',
                'target',
                'enabled',
                'disabled',
                'checked',
                'indeterminate'
            ],
            Formula: ['nth-child', 'nth-last-child', 'nth-of-type', 'nth-last-of-type'],
            Selector: ['not']
        }
    }
});
var selectors4SyntaxDefinition = extendSyntaxDefinition(selectors3SyntaxDefinition, {
    combinators: ['||'],
    attributes: {
        caseSensitivityModifiers: ['i', 'I', 's', 'S']
    },
    pseudoClasses: {
        definitions: {
            NoArgument: [
                'any-link',
                'local-link',
                'target-within',
                'scope',
                'current',
                'past',
                'future',
                'focus-within',
                'focus-visible',
                'read-write',
                'read-only',
                'placeholder-shown',
                'default',
                'valid',
                'invalid',
                'in-range',
                'out-of-range',
                'required',
                'optional',
                'blank',
                'user-invalid'
            ],
            Formula: ['nth-col', 'nth-last-col'],
            String: ['dir'],
            FormulaOfSelector: ['nth-child', 'nth-last-child'],
            Selector: ['current', 'is', 'where', 'has']
        }
    }
});
var progressiveSyntaxDefinition = extendSyntaxDefinition(selectors4SyntaxDefinition, {
    pseudoElements: {
        unknown: 'accept'
    },
    pseudoClasses: {
        unknown: 'accept'
    },
    attributes: {
        unknownCaseSensitivityModifiers: 'accept'
    }
});
exports.cssSyntaxDefinitions = {
    css1: css1SyntaxDefinition,
    css2: css2SyntaxDefinition,
    css3: selectors3SyntaxDefinition,
    'selectors-3': selectors3SyntaxDefinition,
    'selectors-4': selectors4SyntaxDefinition,
    latest: selectors4SyntaxDefinition,
    progressive: progressiveSyntaxDefinition
};
